cranelift-codegen/shared/src/isa/x86/encoding_bits.rs

```diff
@@ -27,61 +27,101 @@ use packed_struct::prelude::*;
 ///        11: 0F 3A <op>  (Op3/Mp3)
 /// 12-14  rrr, opcode bits for the ModR/M byte for certain opcodes.
 /// 15:    REX.W bit (or VEX.W/E)
-#[derive(Copy, Clone, PartialEq, PackedStruct)]
-#[packed_struct(size_bytes = "2", bit_numbering = "lsb0")]
-pub struct EncodingBits {
-    /// Instruction opcode byte, without the prefix.
-    #[packed_field(bits = "0:7")]
-    pub opcode_byte: u8,
-
-    /// Prefix kind for the instruction, as an enum.
-    #[packed_field(bits = "8:11", ty = "enum")]
-    pub prefix: OpcodePrefix,
-
-    /// Bits for the ModR/M byte for certain opcodes.
-    #[packed_field(bits = "12:14")]
-    pub rrr: Integer<u8, packed_bits::Bits3>,
-
-    /// REX.W bit (or VEX.W/E).
-    #[packed_field(bits = "15")]
-    pub rex_w: Integer<u8, packed_bits::Bits1>,
-}
+#[derive(Copy, Clone, PartialEq)]
+pub struct EncodingBits(u16);
+const OPCODE: RangeInclusive<u16> = 0..=7;
+const OPCODE_PREFIX: RangeInclusive<u16> = 8..=11; // Includes pp and mm.
+const RRR: RangeInclusive<u16> = 12..=14;
+const REX_W: RangeInclusive<u16> = 15..=15;
 
 impl From<u16> for EncodingBits {
-    fn from(bits: u16) -> EncodingBits {
-        let bytes: [u8; 2] = [((bits >> 8) & 0xff) as u8, (bits & 0xff) as u8];
-        EncodingBits::unpack(&bytes).expect("failed creating EncodingBits")
+    fn from(bits: u16) -> Self {
+        Self(bits)
     }
 }
 
 impl EncodingBits {
     /// Constructs a new EncodingBits from parts.
     pub fn new(op_bytes: &[u8], rrr: u16, rex_w: u16) -> Self {
-        EncodingBits {
-            opcode_byte: op_bytes[op_bytes.len() - 1],
-            prefix: OpcodePrefix::from_opcode(op_bytes),
-            rrr: (rrr as u8).into(),
-            rex_w: (rex_w as u8).into(),
-        }
+        assert!(
+            op_bytes.len() > 0,
+            "op_bytes must at least include an opcode byte"
+        );
+        let mut new = Self::from(0);
+        let last_byte = op_bytes[op_bytes.len() - 1];
+        new.write(OPCODE, last_byte as u16);
+        let prefix: u8 = OpcodePrefix::from_opcode(op_bytes).into();
+        new.write(OPCODE_PREFIX, prefix as u16);
+        new.write(RRR, rrr);
+        new.write(REX_W, rex_w);
+        new
     }
 
     /// Returns the raw bits.
     #[inline]
     pub fn bits(self) -> u16 {
-        let bytes: [u8; 2] = self.pack();
-        ((bytes[0] as u16) << 8) | (bytes[1] as u16)
+        self.0
+    }
+
+    /// Convenience method for writing bits to specific range.
+    #[inline]
+    fn write(&mut self, range: RangeInclusive<u16>, value: u16) {
+        let size = range.end() - range.start() + 1; // Calculate the number of bits in the range.
+        let mask = (1 << size) - 1; // Generate a bit mask.
+        debug_assert!(
+            value <= mask,
+            "The written value should have fewer than {} bits.",
+            size
```

Comment:
Oh, I didn't realize that assertions had an implicit format! Learned something new :)